controllers/instancegroup_controller.go

```diff
@@ -200,7 +201,9 @@ func (r *InstanceGroupReconciler) Reconcile(ctxt context.Context, req ctrl.Reque
 		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 	}
 
-	r.UpdateStatus(input.InstanceGroup)
+	if !reflect.DeepEqual(existingStatus, input.InstanceGroup.Status) {
```

Comment:
This should not be necessary - when we make an API call to update the resource, if there are no changes, the resource version does not change, and it should not trigger another reconcile in that case.

Are you able to reproduce this?
We are not seeing a scenario where there is an infinite loop

Reply:
I checked in our large clusters with many instance groups and instance-manager seems to be using very little cpu:
```
NAME                                CPU(cores)   MEMORY(bytes)
instance-manager-7f56569896-lp742   4m           26Mi
```

Can you open an issue and possibly add some logs so we can dig into the issue you encountered?

Reply:
I am able to reproduce it - let me see if I can get any more info about what might be going on. Here's some logs for an IG to demonstrate what's going on:

```txt
021-02-08T17:53:19.773Z	INFO	controllers.instancegroup	reconcile event started	{"instancegroup": "instance-manager/arm-nodes", "provisioner": "eks"}
2021-02-08T17:53:19.773Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "Init", "previousState": "Ready"}
2021-02-08T17:53:53.871Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "InitUpdate", "previousState": "Init"}
2021-02-08T17:53:54.073Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "ReconcileModifying", "previousState": "InitUpdate"}
2021-02-08T17:53:54.074Z	INFO	controllers.instancegroup.eks	updated managed policies	{"instancegroup": "arm-nodes", "iamrole": "lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:53:54.074Z	INFO	controllers.instancegroup.eks	reconciled managed role	{"instancegroup": "arm-nodes", "iamrole": "lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:53:54.074Z	INFO	scaling	drift not detected	{"instancegroup": "instance-manager/arm-nodes"}
2021-02-08T17:53:55.072Z	INFO	controllers.instancegroup.eks	bootstrapping arn to aws-auth	{"instancegroup": "arm-nodes", "arn": "arn:aws:iam::<arn>:role/lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:53:55.087Z	INFO	controllers.instancegroup.eks	waiting for node readiness conditions	{"instancegroup": "arm-nodes"}
2021-02-08T17:53:55.087Z	INFO	controllers.instancegroup.eks	desired nodes are ready	{"instancegroup": "arm-nodes", "instances": "i-0524980ffa6b5234b"}
2021-02-08T17:53:55.087Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "ReconcileModified", "previousState": "ReconcileModifying"}
2021-02-08T17:53:55.087Z	INFO	controllers.instancegroup.eks	bootstrapping arn to aws-auth	{"instancegroup": "arm-nodes", "arn": "arn:aws:iam::<arn>:role/lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:53:55.272Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "Ready", "previousState": "ReconcileModified"}
2021-02-08T17:53:55.272Z	INFO	controllers.instancegroup	updating resource status	{"instancegroup": "instance-manager/arm-nodes"}
2021-02-08T17:53:55.290Z	INFO	controllers.instancegroup	reconcile event started	{"instancegroup": "instance-manager/arm-nodes", "provisioner": "eks"}
2021-02-08T17:53:55.290Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "Init", "previousState": "Ready"}
2021-02-08T17:54:31.272Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "InitUpdate", "previousState": "Init"}
2021-02-08T17:54:31.574Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "ReconcileModifying", "previousState": "InitUpdate"}
2021-02-08T17:54:31.977Z	INFO	controllers.instancegroup.eks	updated managed policies	{"instancegroup": "arm-nodes", "iamrole": "lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:54:31.977Z	INFO	controllers.instancegroup.eks	reconciled managed role	{"instancegroup": "arm-nodes", "iamrole": "lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:54:31.977Z	INFO	scaling	drift not detected	{"instancegroup": "instance-manager/arm-nodes"}
2021-02-08T17:54:32.772Z	INFO	controllers.instancegroup.eks	bootstrapping arn to aws-auth	{"instancegroup": "arm-nodes", "arn": "arn:aws:iam::<arn>:role/lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:54:32.972Z	INFO	controllers.instancegroup.eks	waiting for node readiness conditions	{"instancegroup": "arm-nodes"}
2021-02-08T17:54:32.972Z	INFO	controllers.instancegroup.eks	desired nodes are ready	{"instancegroup": "arm-nodes", "instances": "i-0524980ffa6b5234b"}
2021-02-08T17:54:32.972Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "ReconcileModified", "previousState": "ReconcileModifying"}
2021-02-08T17:54:32.972Z	INFO	controllers.instancegroup.eks	bootstrapping arn to aws-auth	{"instancegroup": "arm-nodes", "arn": "arn:aws:iam::<arn>:role/lz-devops-usw2-prd-instance-manager-arm-nodes"}
2021-02-08T17:54:33.172Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "Ready", "previousState": "ReconcileModified"}
2021-02-08T17:54:33.172Z	INFO	controllers.instancegroup	updating resource status	{"instancegroup": "instance-manager/arm-nodes"}
2021-02-08T17:54:33.273Z	INFO	controllers.instancegroup	reconcile event started	{"instancegroup": "instance-manager/arm-nodes", "provisioner": "eks"}
2021-02-08T17:54:33.273Z	INFO	v1alpha1	state transition occured	{"instancegroup": "arm-nodes", "state": "Init", "previousState": "Ready"}

```

Reply:
If you can also provide a redacted CR & version you are using I can also try to reproduce it.

It might be related to a specific change on master branch because we are running latest released version and not seeing this.

Reply:
Yeah this is on latest using LaunchTemplates

Reply:
Here's a CR that is causing it:
```yaml
apiVersion: instancemgr.keikoproj.io/v1alpha1
kind: InstanceGroup
metadata:
  annotations:
    instancemgr.keikoproj.io/cluster-autoscaler-enabled: "true"
    instancemgr.keikoproj.io/os-family: bottlerocket
  labels:
    app.kubernetes.io/managed-by: addonmgr.keikoproj.io
    app.kubernetes.io/name: instance-manager-addon
    app.kubernetes.io/part-of: instance-manager-addon
    app.kubernetes.io/version: v0.0.1
  name: bottlerocket-nodes-zone-1
  namespace: instance-manager
spec:
  eks:
    configuration:
      bootstrapOptions:
        maxPods: 22
      clusterName: lz-devops-usw2-prd
      image: ami-09030b5be1adb0be1
      instanceType: m5.2xlarge
      keyPairName: dev-key-org
      labels:
        instancemgr.keikoproj.io/os-family: bottlerocket
      lifecycleHooks:
      - defaultResult: CONTINUE
        lifecycle: terminate
        name: termination-hook
      managedPolicies:
      - AmazonSSMManagedInstanceCore
      securityGroups:
      - lz-devops-usw2-prd-default
      subnets:
      - NodeASubnet1
      tags:
      - key: k8s.io/cluster-autoscaler/node-template/label/kubernetes.io/arch
        value: amd64
      userData:
      - data: |
          [settings.kernel]
          lockdown = "integrity"
        name: kernel-lsm
        stage: PostBootstrap
      volumes:
      - encrypted: true
        name: /dev/xvda
        size: 2
        type: gp2
      - encrypted: true
        name: /dev/xvdb
        size: 32
        type: gp2
    maxSize: 20
    minSize: 0
    type: LaunchTemplate
  provisioner: eks
  strategy:
    crd:
      crdName: rollingupgrades
      spec: |
        apiVersion: upgrademgr.keikoproj.io/v1alpha1
        kind: RollingUpgrade
        metadata:
          name: "rollup-{{ .InstanceGroup.Name}}"
          namespace: "instance-manager"
        spec:
          postDrainDelaySeconds: 30
          nodeIntervalSeconds: 30
          asgName: {{ .InstanceGroup.Status.ActiveScalingGroupName }}
          strategy:
            mode: eager
            type: randomUpdate
            drainTimeout: -1
      statusFailureString: error
      statusJSONPath: .status.currentStatus
      statusSuccessString: completed
    type: crd
```

Reply:
@backjo still cannot reproduce with a basic launch template CR on latest.
Controller goes idle after ASG is created and nodes are ready:
```
2021-02-08T11:40:14.558-0800	INFO	v1alpha1	state transition occured	{"instancegroup": "hello-world", "state": "ReconcileModified", "previousState": "ReconcileModifying"}
2021-02-08T11:40:14.558-0800	INFO	controllers.instancegroup.eks	bootstrapping arn to aws-auth	{"instancegroup": "hello-world", "arn": "arn:aws:iam::<acct>:role/instancemgr-dev-instance-manager-hello-world"}
2021-02-08T11:40:14.638-0800	INFO	v1alpha1	state transition occured	{"instancegroup": "hello-world", "state": "Ready", "previousState": "ReconcileModified"}
2021-02-08T11:40:14.639-0800	INFO	controllers.instancegroup	updating resource status	{"instancegroup": "instance-manager/hello-world"}
```

It might be a specific feature - if you can provide some version of the CR you are using I can try it out.

Reply:
@backjo 
I was able to submit your CR (with minimal changes to match my cluster) and it ended reconcile.
The only thing I can think of in this case to cause what you are describing is if there is constant scaling going on up/down?
Do you see `desired nodes are ready` in all logs when it's in this state?
Does it keep going forever?
Can you provide larger set of logs (like 10 minutes worth of logs while it's in this state)?

Reply:
Can you also make sure your CRD is up to date with branch you are running? been a while since we made changes, but the last changes were quite large.